### ISSUE TITLE
Parsing empty string should create an error?

### ISSUE THREAD

**Issue by mgeier** [RazrFalcon/xmlparser#8]
First of all, thanks for providing this great crate!

AFAIK, an empty string is not well-formed XML, right?

I think this should print an error (currently it prints nothing):

```rust
for token in xmlparser::Tokenizer::from("") {
    println!("{:?}", token);
}
```

I think the iterator should only yield nothing (without error) when using `enable_fragment_mode()` (because an empty string is a well-formed XML fragment).

**Comment by RazrFalcon** [comment]
I'm not sure, really. Is this stated somewhere in the spec?

**Comment by RazrFalcon** [comment]
`roxmltree` will return an error if there are no root node, so I think that this should be handled on an upper level.

Also, an empty string is a vague term in terms of XML. For example, is this string is empty: `b"\xEF\xBB\xBF \n\t\r &#x20;"`?

**Comment by mgeier** [comment]
Thanks for the quick answer!

I've found https://stackoverflow.com/a/376325/ which links to the relevant section of the spec: https://www.w3.org/TR/REC-xml/#sec-well-formed.

It says that a well-formed XML document "contains one or more elements". Zero elements are not allowed.

> Also, an empty string is a vague term in terms of XML.

I was only thinking about an empty string in the Rust sense, but it turns out that there are many more strings that should be rejected.

> For example, is this string is empty: `b"\xEF\xBB\xBF \n\t\r &#x20;"`?

It is definitely not well-formed and it (correctly) causes an `UnknownToken` error in the `Tokenizer`.

Both `b"\xEF\xBB\xBF"` and `b"&#x20;"` are not allowed at the root level (and both cause an `UnknownToken` error).

This case is more interesting: `b" \n\t\r"`. All those characters are allowed (and they are the only allowed whitespace characters at the root level). But since it doesn't contain an XML element, it is not well-formed XML. The `Tokenizer` iterator doesn't yield anything in this case.

> I think that this should be handled on an upper level.

I'm not sure about that.

I guess it depends on how complicated it is (and how much overhead it causes) to implement this in `xmlparser`. If there happens to be some kind of state already that could be (re-)used, I think it would be a worthwile addition.

Probably in the body of this `if` clause:

https://github.com/RazrFalcon/xmlparser/blob/b586fb73d458187d99ba4f8c763bb23ea541cfc3/src/lib.rs#L419-L421

... there could be an additional check for `!self.fragment_parsing && state != State::AfterElements` which would raise an error. I think the runtime overhead would be negligible, the only problem is that we would need an additional error variant.

The problem with delegating this to the user (= me) is that it can easily be missed. I stumbled upon it more or less by accident. The [limitations](https://docs.rs/xmlparser/0.9.0/xmlparser/index.html#limitations) section in the docs talks about "No tree structure validation", but I wasn't aware that this includes the case of no elements at all.

I guess the correct solution in user code, for now, would be to check that at least one `ElementStart` event has happened.

**Comment by RazrFalcon** [comment]
`\xEF\xBB\xBF` should not produce an error since it's a BOM.

This is a complicated topic. Does XML with DTD but without elements is invalid? Someone may want to parse only DTD. `xmlparser` is a low-level tokenizer, it ignores a lot of checks. For example, duplicated attributes is not an error.

The idea is that you should not use `xmlparser` directly. If you want to parse XML - use `roxmltree`. `xmlparser` is just a tokenizer.

**Comment by mgeier** [comment]
I guess the BOM should be removed by the user when decoding the string (this is apparently done by https://cmcenroe.me/scraper/encoding/types/fn.decode.html, for example). And it should not be fed to the XML parser, right?
AFAIK, the XML spec doesn't know about BOM.

> Does XML with DTD but without elements is invalid?

Yes, it is not well-formed.

> Someone may want to parse only DTD.

OK, I didn't know that anybody would want to do that.

> The idea is that you should not use `xmlparser` directly.

Interesting, you changed the docs two days ago: 117771efb7396e27c289f918ee4dc9771d4ba201

Before that, you were suggesting that `xmlparser` "can be used to write parsers for XML-based formats". And this is exactly my use case. And I'm perfectly satisfied with it! I didn't try `roxmltree`, though, it might be equally suitable for my situation.

**Comment by RazrFalcon** [comment]
>AFAIK, the XML spec doesn't know about BOM.

[It does.](https://www.w3.org/TR/xml/#charencoding)

>Interesting, you changed the docs two days ago

Because it was outdated. `xmlparser` started long before `roxmltree`. If you are parsing a random XML using `xmlparser` and not `roxmltree` you doing it wrong. `roxmltree` does a lot of postprocessing that it impossible on the `xmlparser` level.

In the future, `xmlparser` will probably do even less validation.

**Comment by mgeier** [comment]
Thanks for showing me that the XML spec indeed talks about BOMs.

But I still think what I said above (https://github.com/RazrFalcon/xmlparser/issues/8#issuecomment-526631322) is valid.

The spec says about the BOM:

"This is an encoding signature, not part of either the markup or the character data of the XML document. XML processors MUST be able to use this character to differentiate between UTF-8 and UTF-16 encoded documents."

I would argue that in this case `xmlparser` is only part of the "XML processor". The decoding to a Unicode string is part of the "XML processor", but it has to be done before passing the Unicode string (with the BOM removed!) to `xmlparser`.

> If you are parsing a random XML using `xmlparser` and not `roxmltree` you doing it wrong.

But what if I'm parsing a non-random file containing my own very specific custom XML-based format?
Am I still doing it wrong?

> In the future, `xmlparser` will probably do even less validation.

That's fine for my, I can do some more validation in my own code.

One special problem with this current issue is that there is no way I could get a "span" that I could use to create an error message. As a work-around, I'm currently creating my own `StrSpan` from the whole input string and use that for the error message.

The other problem about overlooking this case is not a big deal anymore since you merged #9.

**Comment by mgeier** [comment]
BTW, I'm using `xmlparser` in this project: https://github.com/AudioSceneDescriptionFormat/asdf-rust

**Comment by RazrFalcon** [comment]
The BOM parsing is just a convenience. The problem is there are no *right* way to parse an XML. At least I'm not aware of it. So I'm parsing in a way that is useful for me.

By the way, xmlparse doesn't check encoding at all. So if Rust's UTF-8 validator accepted the string we assume that it's UTF-8, but it can be a different encoding. This is something that should be investigated, tested and maybe even fixed. Suggestions are welcome.

>Am I still doing it wrong?

No, you probably fine. For the reference, `roxmltree` adds tree validation, namespaces resolving and validation, attributes normalization, checks for duplicated attributes, text unescaping, entity references resolving. And more importantly - a sane API to traverse a tree. (I should update `roxmltree`'s readme...)

>I could get a "span" that I could use to create an error message

Can you clarify this? The whole idea of `StrSpan` is to have a better error messages. If it doesn't fit you needs - feel free to suggest any ideas.

To summarize, I do not use this project directly and do all my XML work via roxmltree. So I don't really care how it works or what checks it has.

**Comment by mgeier** [comment]
Sorry for the late reply ...

Yes, I think it's good to let the user do any text transcoding with Rust, before passing the (then valid UTF-8) string (slice) to `xmlparser`.
But if there was a BOM in the original text data, it should *not* be passed the `xmlparser` (which doesn't have to know what a BOM is).

> No, you're probably fine. For the reference, `roxmltree` adds tree validation, namespaces resolving and validation, attributes normalization, checks for duplicated attributes, text unescaping, entity references resolving.

* *tree validation*: I can do most of this myself, that's no problem. I was just surprised that I have to check manually for a non-existent root element.
* *namespaces*: I don't use those, so I don't need the functionality
* *attributes normalization*: I don't know what that is?
* *duplicated attributes*: I can check those myself, no problem
* *text unescaping*/*entity references resolving*: That would actually be nice to have. What about adding this a a helper function (or multiple functions) to `xmlparser`?

> > I could get a "span" that I could use to create an error message
>
> Can you clarify this? The whole idea of StrSpan is to have a better error messages.

While I'm iterating over `Tokenizer::from(...)` I have access to a `span` field in each token, which I can use to show nice error messages.

But here's the problem: When I reach the end of the iterator and there hasn't been a root element, I don't have a `span`!
Is there a way to get the `span` of the previously processed token (after exiting the `for` loop)?
If there hasn't been a token at all (empty string or only whitespace), there's of course no `span` either.
So how am I supposed to get a `span` in this case?

I would hope that an error could be raised in this case, and I could extract the `span` from this error.

Which leads me to another problem/question:
The tokens provide a `StrSpan` (or multiple), but the error types provide a `TextPos`.
Shouldn't this be the same thing?
I have the feeling that the reason is that `StrSpan` contains a reference while `TextPos` doesn't, but it might be something different.
Why are there two different types to provide essentially the same information?

Also, I think it would actually be nice to have the row and column numbers in `StrSpan`, because right now I have to manually find them by processing the whole input string (and searching for `\n`).

> To summarize, I do not use this project directly and do all my XML work via `roxmltree`. So I don't really care how it works or what checks it has.

OK, I understand.
I guess I won't change my code now (because it's already working), but for future projects `roxmltree` might be the better choice.

**Comment by RazrFalcon** [comment]
I will think about BOM.

>What about adding this a a helper function (or multiple functions) to xmlparser?

You have to have a tree to do this.

>When I reach the end of the iterator and there hasn't been a root element, I don't have a span!

You can just store it in a variable. 

>Shouldn't this be the same thing?

`StrSpan` stores positions in bytes, `TextPos` in chars.

> it would actually be nice to have the row and column numbers in StrSpan

Too expensive.

**Comment by mgeier** [comment]
OK, thanks for the insights!
